src/backend/pipeline/cont_combiner.c

```diff
@@ -812,9 +786,18 @@ read_batch(ContQueryCombinerState *state, TupleBufferBatchReader *reader)
 
 		ExecStoreTuple(heap_copytuple(tbs->tuple->heaptup), state->slot, InvalidBuffer, false);
 		tuplestore_puttupleslot(state->batch, state->slot);
+		state->batch_hashes[count] = tbs->tuple->group_hash;
 
 		IncrementCQRead(1, tbs->size);
 		count++;
+
+		/*
+		 * This is fairly unlikely, but if it does happen we should
+		 * finish processing this batch immediately because we're storing 8-byte hashes
+		 * in memory alongside each of these tuples.
+		 */
+		if (!tuplestore_in_memory(state->batch))
```

Comment:
This is actually not even really related to this PR, but more of a general corner case. The comment doesn't even need to mention the hashes.

Once the initial memory allocated for a tuplestore has been exceeded, new tuples will start spilling to disk. It's better just to finish the batch immediately to avoid that. With the current settings, it would only take 10k ~3KB tuples to cause this to happen, which isn't likely but certainly possible.

I can put this change in a separate diff, like I said I just happened to notice it as I was reading the code.